plonky2/src/plonk/circuit_builder.rs

```diff
@@ -211,7 +206,12 @@ impl<F: RichField + Extendable<D>, const D: usize> CircuitBuilder<F, D> {
     }
 
     /// Adds a gate to the circuit, and returns its index.
-    pub fn add_gate<G: Gate<F, D>>(&mut self, gate_type: G, constants: Vec<F>) -> usize {
+    pub fn add_gate<G: Gate<F, D>>(
+        &mut self,
+        gate_type: G,
+        constants: Vec<F>,
+        params: Vec<F>,
```

Comment:
Could you explain the use case for this?

Reply:
I think I get it now after reading more of the code. Params are like the key used to decide whether two operations can share a gate, correct?

As Nick said elsewhere, I think it could use a comment explaining that, though not sure if it should be here. (As per other comment, maybe we can delete params here and only take it in `find_slot`?)

Reply:
> I think I get it now after reading more of the code. Params are like the key used to decide whether two operations can share a gate, correct?

Yeah exactly. At first I just wanted to use constants to differentiate gates. That works for gates like `ArithmeticGate`, but not for gates like `SwitchGate` which have differentiating parameters that are not in constants. So I added this `params` field which can sometimes be redundant with constants.

> As Nick said elsewhere, I think it could use a comment explaining that, though not sure if it should be here. (As per other comment, maybe we can delete params here and only take it in `find_slot`?)

Sounds good, I'll add comments.